wrappers/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,542 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package pkcs11
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strconv"
+	"sync/atomic"
+
+	uuid "github.com/hashicorp/go-uuid"
+	pkcs11 "github.com/miekg/pkcs11"
+	wrapping "github.com/openbao/go-kms-wrapping/v2"
+)
+
+// These constants contain the accepted env vars; the Vault one is for backwards compat
+const (
+	EnvPkcs11WrapperKeyId   = "PKCS11_WRAPPER_KEY_ID"
+	EnvVaultPkcs11SealKeyId = "VAULT_PKCS11_SEAL_KEY_ID"
+)
+
+// Wrapper is a Wrapper that uses PKCS11
+type Wrapper struct {
+	client       *pkcs11KMS
+	keyId        string
+	currentKeyId *atomic.Value
+}
+
+// Ensure that we are implementing Wrapper
+var _ wrapping.Wrapper = (*Wrapper)(nil)
+
+// NewWrapper creates a new PKCS11 Wrapper
+func NewWrapper() *Wrapper {
+	k := &Wrapper{
+		currentKeyId: new(atomic.Value),
+	}
+	k.currentKeyId.Store("")
+	return k
+}
+
+// SetConfig sets the fields on the Pkcs11Wrapper object based on
+// values from the config parameter.
+//
+// Order of precedence Pkcs11 values:
+// * Environment variable
+// * Value from Vault configuration file
+// * Instance metadata role (access key and secret key)
```

Comment:
I don't think this is true, right? 

Under Azure KMS, we see:

```
        credentialChain := []providers.Provider{
            providers.NewEnvCredentialProvider(),
            providers.NewConfigurationCredentialProvider(credConfig),
            providers.NewInstanceMetadataProvider(),
        }
```

but we don't see that here, because we don't put PKCS#11 pins in the metadata for the running VM (like Azure puts credentials into the VM's metadata that it can access to communicate with other services).